### PR TITLE
fix(server): fix replication master deadlock on cancelation flow

### LINKED ISSUE #3649: Replication deadlock when replica times out
We're seeing a Dragonfly replication deadlock on our test suite, which seems to happen when a replica times out.

I can reproduce the replication deadlock locally (or at least what I'm assuming is the same issue).

Running two Dragonfly processes, but intentionally setting the master process `replication_timeout` to a very small value (100ms) to force the replica to timeout:
```
./dragonfly --alsologtostderr --dbfilename= --port 7000 --replication_timeout 100
./dragonfly --alsologtostderr --dbfilename= --port 7001
```

Then start populating the master with 5GB:
```
redis-cli -p 7000 debug populate 5000000 test 1000 rand
```

Then while its populating configure the replica:
```
redis-cli -p 7001 replicaof localhost 7000
```

The replica will timeout as expected, but then the master is partially deadlocked.

`INFO` hangs even after you shutdown the replica, and attempting to add another replica hangs as the master is unresponsive.

Running v1.22.0 on AWS `t4g.medium`. I added some logs and I think its a deadlock attempting to lock `DflyCmd::mu_` but thats as far as I got

Edit: poking around a bit out of curiosity, `DflyCmd::BreakStalledFlowsInShard` never releases `mu_` as it blocks on `replica_ptr->Cancel()`

### PR DESCRIPTION
**Pull request by adiholden**
fix #3649 
The bug: while running debug populate and cancel replication ,rdb saver cancel is called with context cancel. Both flow calling StopChannel function which sets journal callback to 0 and unregister on journal change and closes channel.
When this flow was called from different fibers than the first call was locked on mutex to unregister the journal the second fiber whould skip the unregister as it was already set the call back id to 0 and it called the close channel leading to deadlock as callback could not push to channel as it was full.
The fiber which consumes from channel already exited because channel was marked as close
The fix:
Now we call close channel only after and if we call the unregister on change, we do this only once and in the cancel function

### CODE REVIEW

File: src/server/rdb_save.cc [hunk 1]
```diff
@@ -1251,15 +1251,7 @@ void RdbSaver::Impl::Cancel() {
...
-  auto& snapshot = GetSnapshot(shard);
-  if (snapshot)
-    snapshot->StopChannel();
-
-  dfly::SliceSnapshot::DbRecord rec;
-  while (channel_.Pop(rec)) {
-  }
-
-  snapshot->Join();
+  GetSnapshot(shard)->Cancel();
```
Comment (romange) [inline src/server/rdb_save.cc:1254]: This function is called from `RdbSaver::Cancel`. Similarly to how we have `RdbSaver::StartSnapshotInShard` and `StartIncrementalSnapshotInShard`, lets rename it to `RdbSaver::CancelInShard`.
Subsequentially, lets dcheck here that that shard is not-null instead checking &  returning.

File: src/server/snapshot.cc [hunk 2]
```diff
@@ -108,32 +108,39 @@ void SliceSnapshot::StartIncremental(Context* cntx, LSN start_lsn) {
...
   journal->UnregisterOnChange(journal_cb_id_);
-  journal_cb_id_ = 0;
+
   PushSerializedToChannel(true);
   CloseRecordChannel();
 }
 
-void SliceSnapshot::StopChannel() {
-  VLOG(1) << "SliceSnapshot::StopChannel";
+// Called only for replication use-case.
```
Comment (romange) [inline src/server/snapshot.cc:130]: Seems that Finalize and Cancel are almost the same. We can unify them into the same function `FinalizeReplication(bool cancel)`.

File: src/server/snapshot.h [hunk 3]
```diff
@@ -180,6 +178,7 @@ class SliceSnapshot {
   // version upper bound for entries that should be saved (not included).
   uint64_t snapshot_version_ = 0;
   uint32_t journal_cb_id_ = 0;
+  bool unregister_journal_ = false;  // true if unregister journal was already called
```
Comment (romange) [inline src/server/snapshot.h:181]: I am still not sure I understand why `journal_cb_id_` can not be used for this.

**Review by romange — COMMENTED** [review]
LGTM!
nit: please add punctuation to the commit description :)

**Review by kostasrim — COMMENTED** [review]
Since this bug is reproducible can you add this test here such that if we ever accidentally make changes around the code of this flow we don't break it ?

**Comment by adiholden** [comment]
> Since this bug is reproducible can you add this test here such that if we ever accidentally make changes around the code of this flow we don't break it ?

yes I am working on this

**Comment by adiholden** [comment]
Created a seprate PR for testing- https://github.com/dragonflydb/dragonfly/pull/3691
It fails on master and passes on this bransh

**Review by romange — APPROVED** [review]
Less lines of code and less bugs. A perfect PR!

**Pull request merged**